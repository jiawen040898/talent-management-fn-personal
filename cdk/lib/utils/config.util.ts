import type { Construct } from 'constructs';

import * as productionEnvVariables from '../configs/production.config';
import * as sandboxEnvVariables from '../configs/sandbox.config';
import * as stagingEnvVariables from '../configs/staging.config';
import { CDKEnvironmentVariables } from '../interfaces';

const getEnvironmentVariables = (
    scope: Construct,
    environment: string,
): CDKEnvironmentVariables => {
    const config = {
        sandbox: sandboxEnvVariables.config(scope),
        staging: stagingEnvVariables.config(scope),
        production: productionEnvVariables.config(scope),
    };

    return config[environment as keyof typeof config];
};

export const configUtil = {
    getEnvironmentVariables,
};
