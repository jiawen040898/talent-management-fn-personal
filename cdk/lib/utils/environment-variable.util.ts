import type { Construct } from 'constructs';

import * as productionEnvVariables from '../environment-variables/production-env-variables';
import * as sandboxEnvVariables from '../environment-variables/sandbox-env-variables';
import * as stagingEnvVariables from '../environment-variables/staging-env-variables';

const getEnvironmentVariables = (scope: Construct, environment: string) => {
    const config = {
        sandbox: sandboxEnvVariables.envVariables(scope),
        staging: stagingEnvVariables.envVariables(scope),
        production: productionEnvVariables.envVariables(scope),
    };

    return config[environment as keyof typeof config];
};

export const environmentVariableUtil = {
    getEnvironmentVariables,
};
