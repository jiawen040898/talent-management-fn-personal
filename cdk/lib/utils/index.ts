export * from './config.util';
export * from './esbuild/esbuild-bundling.util';
