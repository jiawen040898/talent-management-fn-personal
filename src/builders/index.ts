export * from './base-query.builder';
export * from './employee-task-query.builder';
export * from './feedback-cycle-reviewee-query.builder';
export * from './feedback-cycle-reviewer-query.builder';
export * from './learning-goal-query.builder';
