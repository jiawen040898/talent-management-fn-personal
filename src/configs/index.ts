export * from './aws.config';
export * from './gcp.config';
export * from './recommendation.config';
