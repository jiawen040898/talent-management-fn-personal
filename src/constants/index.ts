export * from './action-source.enum';
export * from './big-queries.enum';
export * from './cache-store.enum';
export * from './company-module.enum';
export * from './company-status.enum';
export * from './due-type.enum';
export * from './employee-score.enum';
export * from './employee-task.enum';
export * from './event-type.enum';
export * from './feedback-recommendation.enum';
export * from './fit-score-recipe.enum';
export * from './person-score.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework-type.enum';
export * from './sqs-event-type.enum';
export * from './unleash-errors.enum';
