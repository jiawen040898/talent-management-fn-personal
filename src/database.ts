import { secretService } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import { ENTITIES } from './models';

type DbSecret = {
    username: string;
    engine: string;
    port: number;
    host: string;
    password: string;
    schema: string;
    dbname: string;
    dbInstanceIdentifier: string;
};

let dataSource: DataSource;

export const getDataSource = async (): Promise<DataSource> => {
    if (dataSource?.isInitialized) {
        return dataSource;
    }

    const dbSecret = await secretService.getSecret<DbSecret>(
        process.env.SM_NAME as string,
    );

    dataSource = await new DataSource({
        type: 'postgres',
        host: dbSecret.host,
        port: dbSecret.port,
        username: dbSecret.username,
        password: dbSecret.password,
        database: dbSecret.dbname,
        schema: dbSecret.schema,
        entities: [...ENTITIES],
        maxQueryExecutionTime: 1000,
        synchronize: false,
        migrationsRun: false,
        migrations: [],
        extra: {
            connectionTimeoutMillis: 3000,
            idleTimeoutMillis: 1000,
            max: 1,
        },
        logging: ['error'],
    }).initialize();

    return dataSource;
};
