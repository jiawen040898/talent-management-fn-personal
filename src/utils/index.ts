export * from './company.util';
export * from './date.util';
export * from './number.util';
export * from './path.util';
export * from './url-generator.util';
