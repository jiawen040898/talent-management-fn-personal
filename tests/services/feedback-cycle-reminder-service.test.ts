import { dateTimeUtil, snsService } from '@pulsifi/fn';
import { DataSource } from 'typeorm';

import * as database from '../../src/database';
import { feedbackCycleReminderService } from '../../src/services/feedback-cycle-reminder.service';
import { dateUtil } from '../../src/utils';
import { getTestDataSourceAndAddData } from '../setup';
import { testData } from './fixtures/test-data';

jest.mock('@pulsifi/fn/services/sns.service');

describe('FeedbackCycleReminderService', () => {
    let dataSource: DataSource;
    const spygetDataSource = jest.spyOn(database, 'getDataSource');
    const urlMatchedRegex = '^https://local-employee.pulsifi.me/';

    beforeAll(async () => {
        dataSource = await getTestDataSourceAndAddData(
            testData.entitiesFeedbackCycleReminderToBeAdded,
        );
        spygetDataSource.mockImplementation(() => {
            return Promise.resolve(dataSource);
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('send nomination submission reminder with x day before deadline to reviewee', async () => {
        // Arrange
        jest.mocked(snsService.sendEventModel).mockResolvedValueOnce();

        // Act
        await feedbackCycleReminderService.sendNominationSubmissionReminder({
            ...testData.mockCompanyLookupData,
            localDate: dateUtil.getDate(
                dateTimeUtil.addDays(1, new Date()).toISOString(),
            ),
        });

        // Assert
        expect(
            jest.mocked(snsService.sendEventModel).mock.calls[0][0],
        ).toMatchSnapshot({
            event_id: expect.anything(),
            data: {
                cycle: {
                    start_date: expect.anything(),
                    close_date: expect.anything(),
                    nomination_submission_deadline: expect.anything(),
                    nomination_approval_deadline: expect.anything(),
                    feedback_submission_deadline: expect.anything(),
                },
                call_to_action: {
                    url: expect.stringMatching(urlMatchedRegex),
                },
            },
        });
    });

    it('send nomination approval reminder with x day before deadline to reporting manager of reviewee', async () => {
        // Arrange
        jest.mocked(snsService.sendEventModel).mockResolvedValueOnce();

        // Act
        await feedbackCycleReminderService.sendManagerNominationApprovalReminder(
            {
                ...testData.mockCompanyLookupData,
                localDate: dateUtil.getDate(
                    dateTimeUtil.addDays(4, new Date()).toISOString(),
                ),
            },
        );

        // Assert
        expect(
            jest.mocked(snsService.sendEventModel).mock.calls[0][0],
        ).toMatchSnapshot({
            event_id: expect.anything(),
            data: {
                cycle: {
                    start_date: expect.anything(),
                    close_date: expect.anything(),
                    nomination_submission_deadline: expect.anything(),
                    nomination_approval_deadline: expect.anything(),
                    feedback_submission_deadline: expect.anything(),
                },

                call_to_action: {
                    url: expect.stringMatching(urlMatchedRegex),
                },
            },
        });
    });

    it('send feedback submission reminder with x day before deadline to reviewer (self | others)', async () => {
        // Arrange
        jest.mocked(snsService.sendEventModel).mockClear();

        // Act
        await feedbackCycleReminderService.sendFeedbackSubmissionReminder({
            ...testData.mockCompanyLookupData,
            localDate: dateUtil.getDate(
                dateTimeUtil.addDays(7, new Date()).toISOString(),
            ),
        });

        // Assert
        const expectDataSnapShot = {
            cycle: {
                id: expect.anything(),
                start_date: expect.anything(),
                close_date: expect.anything(),
                nomination_submission_deadline: expect.anything(),
                nomination_approval_deadline: expect.anything(),
                feedback_submission_deadline: expect.anything(),
            },
            call_to_action: {
                url: expect.stringMatching(urlMatchedRegex),
            },
        };

        expect(
            jest.mocked(snsService.sendEventModel).mock.calls[0][0],
        ).toMatchSnapshot({
            event_id: expect.anything(),
            data: expectDataSnapShot,
        });
        expect(
            jest.mocked(snsService.sendEventModel).mock.calls[1][0],
        ).toMatchSnapshot({
            event_id: expect.anything(),
            data: expectDataSnapShot,
        });
        expect(
            jest.mocked(snsService.sendEventModel).mock.calls[2][0],
        ).toMatchSnapshot({
            event_id: expect.anything(),
            data: expectDataSnapShot,
        });
    });
});
