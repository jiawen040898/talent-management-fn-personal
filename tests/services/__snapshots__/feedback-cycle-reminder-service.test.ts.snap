// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackCycleReminderService send feedback submission reminder with x day before deadline to reviewer (self | others) 1`] = `
{
  "company_id": 5,
  "data": {
    "call_to_action": {
      "url": StringMatching /\\^https:\\\\/\\\\/local-employee\\.pulsifi\\.me\\\\//,
    },
    "company": {
      "name": "Bitcoin Incorporation",
    },
    "cycle": {
      "close_date": Anything,
      "feedback_submission_deadline": Anything,
      "id": Anything,
      "name": "Cycle Name 3",
      "nomination_approval_deadline": Anything,
      "nomination_submission_deadline": Anything,
      "start_date": Anything,
    },
    "email_recipient": {
      "email": "external@yahoo.com",
      "name": "External 1",
    },
    "email_sender": {
      "name": "Bitcoin Incorporation",
    },
    "email_template": {
      "body_in_html": "<p>Hello, this is a feedback reminder for benny </p>",
      "subject": "Please complete feedback for Benny",
    },
    "provider": {
      "email": "external@yahoo.com",
      "employee_id": null,
      "first_name": "External 1",
      "full_name": "External 1",
      "id": "00000000-0000-0000-0000-000000000014",
      "last_name": "External 1",
      "user_account_id": 0,
    },
    "receiver": {
      "email": "employee17@test.pulsifi.me",
      "employee_id": "00000000-0000-0000-0000-000000000002",
      "first_name": "First Name 17",
      "full_name": "First Name 17 Last Name 17",
      "id": "00000000-0000-0000-0000-000000000009",
      "last_name": "Last Name 17",
      "user_account_id": 117,
    },
  },
  "event_id": Anything,
  "event_type": "talent_management_feedback_reviewer_response_submission_due",
  "user_account_id": 0,
}
`;

exports[`FeedbackCycleReminderService send feedback submission reminder with x day before deadline to reviewer (self | others) 2`] = `
{
  "company_id": 5,
  "data": {
    "call_to_action": {
      "url": StringMatching /\\^https:\\\\/\\\\/local-employee\\.pulsifi\\.me\\\\//,
    },
    "company": {
      "name": "Bitcoin Incorporation",
    },
    "cycle": {
      "close_date": Anything,
      "feedback_submission_deadline": Anything,
      "id": Anything,
      "name": "Cycle Name 3",
      "nomination_approval_deadline": Anything,
      "nomination_submission_deadline": Anything,
      "start_date": Anything,
    },
    "email_recipient": {
      "email": "employee21@test.pulsifi.me",
      "name": "First Name 21 Last Name 21",
    },
    "email_sender": {
      "name": "Bitcoin Incorporation",
    },
    "email_template": {
      "body_in_html": "<p>Hello, this is a feedback reminder </p>",
      "subject": "Please complete your feedback",
    },
    "provider": {
      "email": "employee21@test.pulsifi.me",
      "employee_id": "00000000-0000-0000-0000-000000000006",
      "first_name": "First Name 21",
      "full_name": "First Name 21 Last Name 21",
      "id": "00000000-0000-0000-0000-000000000012",
      "last_name": "Last Name 21",
      "user_account_id": 121,
    },
    "receiver": {
      "email": "employee17@test.pulsifi.me",
      "employee_id": "00000000-0000-0000-0000-000000000002",
      "first_name": "First Name 17",
      "full_name": "First Name 17 Last Name 17",
      "id": "00000000-0000-0000-0000-000000000009",
      "last_name": "Last Name 17",
      "user_account_id": 117,
    },
  },
  "event_id": Anything,
  "event_type": "talent_management_feedback_reviewee_response_submission_due",
  "user_account_id": 121,
}
`;

exports[`FeedbackCycleReminderService send feedback submission reminder with x day before deadline to reviewer (self | others) 3`] = `
{
  "company_id": 5,
  "data": {
    "call_to_action": {
      "url": StringMatching /\\^https:\\\\/\\\\/local-employee\\.pulsifi\\.me\\\\//,
    },
    "company": {
      "name": "Bitcoin Incorporation",
    },
    "cycle": {
      "close_date": Anything,
      "feedback_submission_deadline": Anything,
      "id": Anything,
      "name": "Cycle Name 3",
      "nomination_approval_deadline": Anything,
      "nomination_submission_deadline": Anything,
      "start_date": Anything,
    },
    "email_recipient": {
      "email": "employee22@test.pulsifi.me",
      "name": "First Name 22 Last Name 22",
    },
    "email_sender": {
      "name": "Bitcoin Incorporation",
    },
    "email_template": {
      "body_in_html": "<p>Hello, this is a feedback reminder for benny </p>",
      "subject": "Please complete feedback for Benny",
    },
    "provider": {
      "email": "employee22@test.pulsifi.me",
      "employee_id": "00000000-0000-0000-0000-000000000007",
      "first_name": "First Name 22",
      "full_name": "First Name 22 Last Name 22",
      "id": "00000000-0000-0000-0000-000000000013",
      "last_name": "Last Name 22",
      "user_account_id": 122,
    },
    "receiver": {
      "email": "employee17@test.pulsifi.me",
      "employee_id": "00000000-0000-0000-0000-000000000002",
      "first_name": "First Name 17",
      "full_name": "First Name 17 Last Name 17",
      "id": "00000000-0000-0000-0000-000000000009",
      "last_name": "Last Name 17",
      "user_account_id": 117,
    },
  },
  "event_id": Anything,
  "event_type": "talent_management_feedback_reviewer_response_submission_due",
  "user_account_id": 122,
}
`;

exports[`FeedbackCycleReminderService send nomination approval reminder with x day before deadline to reporting manager of reviewee 1`] = `
{
  "company_id": 5,
  "data": {
    "call_to_action": {
      "url": StringMatching /\\^https:\\\\/\\\\/local-employee\\.pulsifi\\.me\\\\//,
    },
    "company": {
      "name": "Bitcoin Incorporation",
    },
    "cycle": {
      "close_date": Anything,
      "feedback_submission_deadline": Anything,
      "id": "00000000-0000-0000-0000-000000000015",
      "name": "Cycle Name 3",
      "nomination_approval_deadline": Anything,
      "nomination_submission_deadline": Anything,
      "start_date": Anything,
    },
    "email_recipient": {
      "email": "employee18@test.pulsifi.me",
      "name": "First Name 18 Last Name 18",
    },
    "email_sender": {
      "name": "Bitcoin Incorporation",
    },
    "manager": {
      "email": "employee18@test.pulsifi.me",
      "employee_id": "00000000-0000-0000-0000-000000000003",
      "first_name": "First Name 18",
      "full_name": "First Name 18 Last Name 18",
      "last_name": "Last Name 18",
      "user_account_id": 118,
    },
    "receiver": {
      "email": "employee17@test.pulsifi.me",
      "employee_id": "00000000-0000-0000-0000-000000000002",
      "first_name": "First Name 17",
      "full_name": "First Name 17 Last Name 17",
      "id": "00000000-0000-0000-0000-000000000009",
      "last_name": "Last Name 17",
      "user_account_id": 117,
    },
    "total_nomination": 1,
  },
  "event_id": Anything,
  "event_type": "talent_management_feedback_manager_nomination_approval_due",
  "user_account_id": 118,
}
`;

exports[`FeedbackCycleReminderService send nomination submission reminder with x day before deadline to reviewee 1`] = `
{
  "company_id": 5,
  "data": {
    "call_to_action": {
      "url": StringMatching /\\^https:\\\\/\\\\/local-employee\\.pulsifi\\.me\\\\//,
    },
    "company": {
      "name": "Bitcoin Incorporation",
    },
    "cycle": {
      "close_date": Anything,
      "feedback_submission_deadline": Anything,
      "id": "00000000-0000-0000-0000-000000000015",
      "name": "Cycle Name 3",
      "nomination_approval_deadline": Anything,
      "nomination_submission_deadline": Anything,
      "start_date": Anything,
    },
    "email_recipient": {
      "email": "employee16@test.pulsifi.me",
      "name": "First Name 16 Last Name 16",
    },
    "email_sender": {
      "name": "Bitcoin Incorporation",
    },
    "manager": {
      "email": null,
      "employee_id": null,
      "first_name": null,
      "full_name": "null",
      "last_name": null,
      "user_account_id": null,
    },
    "receiver": {
      "email": "employee16@test.pulsifi.me",
      "employee_id": "00000000-0000-0000-0000-000000000001",
      "first_name": "First Name 16",
      "full_name": "First Name 16 Last Name 16",
      "id": "00000000-0000-0000-0000-000000000008",
      "last_name": "Last Name 16",
      "user_account_id": 116,
    },
    "total_nomination": 0,
  },
  "event_id": Anything,
  "event_type": "talent_management_feedback_nomination_submission_due",
  "user_account_id": 116,
}
`;
